### PR TITLE
fix(redirects): handle non-verbatim targets

### LINKED ISSUE #8914: Configured Redirects with more than two params don't expand properly
### Astro Info

```block
Astro                    v3.3.2
Node                     v18.17.1
System                   Linux (x64)
Package Manager          pnpm
Output                   server
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind
                         @astrojs/mdx
                         @astrojs/react
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

#### Context
Redirects configuration in `astro.config.mjs`

Tested usint the following configuration (however certainly not limited to this configuration):
- SSR mode 
- Node adapter

#### Bug
Redirects entries with more than one param (ie: '/ov/[page]/[...user]': '/tests/ov/[page]/[...user]') don't  work.
The expansion does not take place.

#### Example:

Given the following configuration:

```javascript
export default defineConfig({
  redirects: {
    '/ov/[page]/[...user]': '/tests/ov/[page]/[...user]', // does not work
    '/ot/[...user]': '/tests/ot/[...user]',               // works
  },
  integrations: [tailwind(), mdx(), react()],
  output: "server",
  adapter: node({
    mode: "standalone",
  }),
})
```

- http://localhost:4321/ot/some-page/some-user/test 
will redirect properly to
- http://localhost:4321/tests/ot/some-page/some-user/test

**While**
- http://localhost:4321/ov/some-page/some-user/test
will redirect without expanding the params to
- http://localhost:4321/tests/ov/[page]/[...user]

### What's the expected result?

Given the configuration above,

- http://localhost:4321/ov/some-page/some-user/test
Should redirect to 
- http://localhost:4321/tests/ov/some-page/some-user/test
Instead of
- http://localhost:4321/tests/ov/[page]/[...user]

### Link to Minimal Reproducible Example
https://github.com/lilnasy/redirect-issue

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by lilnasy**
## Changes
- Fixes #8914

## Testing
- Added 3 cases to [redirects.test.js](https://github.com/withastro/astro/pull/9089/files#diff-f0e8bf856624a391d496c98638e77e91ca1f9ad3a3c30d4d14ff722ff6e6da97)

## Docs
Does not affect usage.

### CODE REVIEW

File: packages/astro/src/core/redirects/helpers.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export function redirectRouteGenerate(redirectRoute: RouteData, data: Params): s
 	if (typeof routeData !== 'undefined') {
 		return routeData?.generate(data) || routeData?.pathname || '/';
 	} else if (typeof route === 'string') {
+		console.log({ data, pattern: route});
```
Comment (lilnasy) [inline packages/astro/src/core/redirects/helpers.ts:23]: image

**Comment by florian-lefebvre** [comment]
@lilnasy what's the status of this PR? Did you encounter some blocker?

**Comment by lilnasy** [comment]
It's just here to explain the code path that goes wrong.

File: packages/astro/test/redirects.test.js [hunk 2]
```diff
@@ -68,10 +75,31 @@ describe('Astro.redirect', () => {
...
+				const app = await fixture.loadTestAdapterApp();
+				const request = new Request('http://example.com/source/x/y/z');
+				const response = await app.render(request);
+				expect(response.headers.get("Location")).to.equal("/not-verbatim/target3/x/y/z");
+			});
 		});
 	});
 
-	describe('output: "static"', () => {
+	describe.skip('output: "static"', () => {
```
Comment (ematipico) [inline packages/astro/test/redirects.test.js:102]: Is it intended? If so, could you add a comment or `TODO` to tell us when we should unskip it?

File: packages/astro/test/redirects.test.js [hunk 3]
```diff
@@ -217,7 +245,7 @@ describe('Astro.redirect', () => {
 		});
 	});
 
-	describe('config.build.redirects = false', () => {
+	describe.skip('config.build.redirects = false', () => {
```
Comment (ematipico) [inline packages/astro/test/redirects.test.js:248]: Is it intended? If so, could you add a comment or `TODO` to tell us when we should unskip it?

File: packages/astro/src/core/redirects/helpers.ts [hunk 4]
```diff
@@ -20,6 +20,16 @@ export function redirectRouteGenerate(redirectRoute: RouteData, data: Params): s
 	if (typeof routeData !== 'undefined') {
 		return routeData?.generate(data) || routeData?.pathname || '/';
 	} else if (typeof route === 'string') {
+		// TODO: this logic is duplicated between here and manifest/create.ts
+		if (Object.keys(data).length > 0) {
+			let target = route
+			for (const param in data) {
+				const paramValue = data[param]!
```
Comment (ematipico) [inline packages/astro/src/core/redirects/helpers.ts:27]: nit: I don't like very much the use of `for..in`, we should use `for..of`

**Review by ematipico — APPROVED** [review]

File: packages/astro/src/core/redirects/helpers.ts (hunk 4 above)
Comment (lilnasy) [reply packages/astro/src/core/redirects/helpers.ts:27]: its not an array

File: packages/astro/test/redirects.test.js (hunk 3 above)
Comment (lilnasy) [reply packages/astro/test/redirects.test.js]: good catch, this was just for testing locally, unskipped

File: packages/astro/src/core/redirects/helpers.ts (hunk 4 above)
Comment (ematipico) [reply packages/astro/src/core/redirects/helpers.ts]:
```js
const dataAsArray  =Object.entries(data);

if (dataAsArray.length >) {
	for..of
}
```

Comment (lilnasy) [reply packages/astro/src/core/redirects/helpers.ts]: oh didnt see this, implemented using object.keys

**Review by natemoo-re — APPROVED** [review]
LGTM! Thanks for the fix

**Pull request merged**